### ISSUE TITLE
Can I wrap words for long column content?

### ISSUE THREAD

**Issue by leoc** [matt-harvey/tabulo#12]
I could not find any information in the docs or I am blind right now :)

Is it possible to let the column wrap before the word that is too long for the column?

So instead of:

```
+------------+
| Test       |
+------------+
| This is lo |
| ng         | 
+------------+
```

I would see:

```
+------------+
| Test       |
+------------+
| This is    |
| long       | 
+------------+
```

Cheerio!

![](https://media2.giphy.com/media/qngUbAbaWRuFy/giphy.gif)

**Comment by matt-harvey** [comment]
Hi @leoc,

At the moment, this can only do be done manually: If the [callable](https://github.com/matt-harvey/tabulo#full-api-) passed to `Tabulo::Table#add_column` returns a string containing a line break (`\n` or '\r\n', depending on your OS), then the cell will always wrap at that specific point (assuming the `formatter:` option is not overridden). Alternatively this could be done in the [formatter](https://github.com/matt-harvey/tabulo#formatting-cell-values-).

This can't be done automatically, however. The automatic wrapping feature always wraps at a character level (or strictly speaking, at a Unicode grapheme cluster level), not a word level. (See the `Tabulo::Cell#calculate_subcells` method for details.)

Automatic wrapping at word boundaries is a feature that I'd love Tabulo to have. This could take the form of a `wrap_at:` parameter passed to `#add_column`, taking a value of either `:word` or `:grapheme_cluster` (which would be the default).

I'm not sure when I'm going to have time to work on it though. If you have the time and inclination, a PR to add this feature would be very welcome! :)

In any case, I'll leave this Issue open as a feature request. Thanks for raising it!

**Comment by leoc** [comment]
I will see if I can find a few minutes in the upcoming week. Thanks for the hints <3

![](https://media4.giphy.com/media/l0HTYUmU67pLWv1a8/giphy.gif)

**Comment by matt-harvey** [comment]
Awesome, thanks @leoc :smile:

**Comment by matt-harvey** [comment]
Just to keep this thread alive... This feature is still on my radar, and I would definitely like to see it in Tabulo. I've just been very busy with other things these past few weeks, and haven't been able to find the time to work on this myself.

If anyone (@leoc ? someone else?) would like to have a go at adding this feature, then that would be hugely appreciated.

If you're reading and are interested in contributing, please see above for pointers on where in the codebase to start. The solution probably involves doing a `string.split(/\W+/)` and then manipulating the results to optimally fit the column's width. I suggest writing tests first and then writing code to make them pass; in my experience, developing Tabulo is particularly well-suited to a TDD approach.

Having said that, if noone else is able to work on this, then I'll probably get on it myself in the next few weeks when I finally do find the time.

**Comment by matt-harvey** [comment]
I have now implemented this feature. Released in [v2.7.0](https://github.com/matt-harvey/tabulo/releases/tag/v2.7.0).

See the [README](https://github.com/matt-harvey/tabulo#preserve-words) for details.

**Comment by leoc** [comment]
Oh my, I totally forgot looking into it again. Had a lot on my plate recently. Thanks so much for implementing that feature! :+1:
